servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1478,6 +1519,185 @@ public void setState(long fileId, SetStateOptions options)
+      inode.setPersistenceState(PersistenceState.IN_PROGRESS);
+    }
+
+    synchronized (mWorkerToAsyncPersistFiles) {
+      if (!mWorkerToAsyncPersistFiles.containsKey(workerId)) {
+        mWorkerToAsyncPersistFiles.put(workerId, Sets.<Long>newHashSet());
+      }
+      mWorkerToAsyncPersistFiles.get(workerId).add(fileId);
+    }
+
+    return workerId;
+  }
+
+  /**
+   * Gets a worker where the given file is stored.
+   *
+   * @param fileId the file id, -1 if no worker can be found
+   * @return the storing worker
+   * @throws FileDoesNotExistException when the file does not exist on any worker
+   */
+  private long getWorkerStoringFile(long fileId) throws FileDoesNotExistException {
+    Map<Long, Integer> workerBlockCounts = Maps.newHashMap();
+    int totalBlockNum;
+    try {
+      totalBlockNum = getFileBlockInfoList(fileId).size();
+
+      for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+        for (BlockLocation blockLocation : fileBlockInfo.blockInfo.locations) {
+          if (workerBlockCounts.containsKey(blockLocation.workerId)) {
+            workerBlockCounts.put(blockLocation.workerId,
+                workerBlockCounts.get(blockLocation.workerId) + 1);
+          } else {
+            workerBlockCounts.put(blockLocation.workerId, 1);
+          }
+        }
+      }
+    } catch (FileDoesNotExistException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    } catch (InvalidPathException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    }
+
+    if (workerBlockCounts.size() == 0) {
+      LOG.error("The file " + fileId + " does not exist on any worker");
+      return IdUtils.INVALID_WORKER_ID;
+    } else if (workerBlockCounts.size() > 1) {
+      LOG.info("The file is stored at more than one worker: " + workerBlockCounts);
+    }
+
+    // return the first worker that has all the blocks
+    for (Entry<Long, Integer> entry : workerBlockCounts.entrySet()) {
+      if (entry.getValue() == totalBlockNum) {
+        return entry.getKey();
+      }
+    }
+
+    LOG.error("Not all the blocks of file {} stored on the same worker", fileId);
+    return IdUtils.INVALID_WORKER_ID;
+  }
+
+  /**
+   * Polls the files to send to the given worker for persistence.
+   *
+   * @param workerId the worker id
+   * @return the list of files
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws InvalidPathException if the path is invalid
+   */
+  private List<PersistFile> pollToCheckpoint(long workerId)
+      throws FileDoesNotExistException, InvalidPathException {
+    List<PersistFile> files = Lists.newArrayList();
+    synchronized (mWorkerToAsyncPersistFiles) {
+      if (!mWorkerToAsyncPersistFiles.containsKey(workerId)) {
+        return files;
+      }
+    }
+
+    List<Long> toRequestFilePersistence = Lists.newArrayList();
```

Comment:
rename to `fileIdsToPersist`? Is this just the fileIds for the `PersistFile`s in `files`? If so we should group them together, have a comment about this invariant, and name them more similarly